tutoraspects/templates/aspects/apps/aspects/migrations/alembic/versions/0009_prefer_course_key.py

```diff
@@ -0,0 +1,213 @@
+        SELECT
+            event_id,
+            emission_time,
+            actor_id,
+            object_id,
+            course_id,
+            org,
+            verb_id,
+            JSON_VALUE(event_str, '$.result.response') as responses,
+            JSON_VALUE(event_str, '$.result.score.scaled') as scaled_score,
+            if(
+                verb_id = 'https://w3id.org/xapi/acrossx/verbs/evaluated',
+                cast(JSON_VALUE(event_str, '$.result.success') as Bool),
+                false
+            ) as success,
+            JSON_VALUE(event_str, '$.object.definition.interactionType') as interaction_type,
+            if(
+                verb_id = 'https://w3id.org/xapi/acrossx/verbs/evaluated',
+                cast(JSON_VALUE(event_str, '$.object.definition.extensions."http://id.tincanapi.com/extension/attempt-id"') as Int16),
+                0
+            ) as attempts
+        FROM
+            {{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_XAPI_TABLE }}
+        WHERE
+            verb_id in (
+                'https://w3id.org/xapi/acrossx/verbs/evaluated',
+                'http://adlnet.gov/expapi/verbs/passed',
+                'http://adlnet.gov/expapi/verbs/asked'
+            );
+        """,
+    "{{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_NAVIGATION_TRANSFORM_MV }}": """
+        CREATE MATERIALIZED VIEW IF NOT EXISTS {{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_NAVIGATION_TRANSFORM_MV }}
+        TO {{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_NAVIGATION_EVENTS_TABLE }} AS
+        SELECT
+            event_id,
+            emission_time,
+            actor_id,
+            object_id,
+            course_id,
+            org,
+            verb_id,
+            JSON_VALUE(event_str, '$.object.definition.type') AS object_type,
+            -- clicking a link and selecting a module outline have no starting-position field
+            if (
+                object_type in (
+                    'http://adlnet.gov/expapi/activities/link',
+                    'http://adlnet.gov/expapi/activities/module'
+                ),
+                0,
+                cast(JSON_VALUE(
+                    event_str,
+                    '$.context.extensions."http://id.tincanapi.com/extension/starting-position"'
+                ) as Int16)
+            ) AS starting_position,
+            JSON_VALUE(
+                event_str,
+                '$.context.extensions."http://id.tincanapi.com/extension/ending-point"'
+            ) AS ending_point
+        FROM
+            {{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_XAPI_TABLE }}
+        WHERE verb_id IN (
+            'https://w3id.org/xapi/dod-isd/verbs/navigated'
+        );
+        """,
+}
+
+
+def upgrade():
+    # drop the tables and MVs and create new ones, with course_id
+    # replaced by course_key
+    for name, ddl in TABLES.items():
+        op.execute(f"DROP TABLE IF EXISTS {name}")
+        op.execute(ddl.replace("course_id", "course_key"))
```

Comment:
Is there a reason not to just do this in the DDL?

Reply:
The only reason was to avoid having two copies of every query and DDL statement in the file, which I admit isn't really a great reason. I tried pulling the queries and DDL statements out into variables in the other migration files so I could import them here, but the Alembic file naming convention results in invalid module names (unless we did `importlib.import_module("0003_enrollment")`, which feels a little too awkward).